vault/external_tests/plugin/plugin_test.go

```diff
@@ -0,0 +1,1073 @@
+			resp, err := core.HandleRequest(namespace.RootContext(testCtx), req)
+			if err != nil {
+				t.Fatalf("err: %v", err)
+			}
+			if resp == nil {
+				t.Fatalf("bad: response should not be nil")
+			}
+
+			// Seal the cluster
+			cluster.EnsureCoresSealed(t)
+
+			// Simulate removal of the plugin binary. Use os.Args to determine file name
+			// since that's how we create the file for catalog registration in the test
+			// helper.
+			pluginFileName := filepath.Base(os.Args[0])
+			err = os.Remove(filepath.Join(cluster.TempDir, pluginFileName))
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// Unseal the cluster
+			cluster.UnsealCores(t)
+
+			// Make a request against on tune after it is removed
+			req = logical.TestRequest(t, logical.ReadOperation, "sys/mounts/mock-0/tune")
+			req.ClientToken = core.Client.Token()
+			_, err = core.HandleRequest(namespace.RootContext(testCtx), req)
+			if err == nil {
+				t.Fatalf("expected error")
+			}
+		})
+	}
+}
+
+func TestSystemBackend_Plugin_MismatchType(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		pluginVersion string
+	}{
+		{
+			pluginVersion: "v5_multiplexed",
+		},
+		{
+			pluginVersion: "v5",
+		},
+		{
+			pluginVersion: "v4",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.pluginVersion, func(t *testing.T) {
+			cluster := testSystemBackendMock(t, 1, 1, logical.TypeLogical, tc.pluginVersion)
+			defer cluster.Cleanup()
+
+			core := cluster.Cores[0]
+
+			// Add a credential backend with the same name
+			vault.TestAddTestPlugin(t, core.Core, "mock-plugin", consts.PluginTypeCredential, "", "TestBackend_PluginMainCredentials", []string{}, "")
+
+			// Make a request to lazy load the now-credential plugin
+			// and expect an error
+			req := logical.TestRequest(t, logical.ReadOperation, "mock-0/internal")
+			req.ClientToken = core.Client.Token()
+			_, err := core.HandleRequest(namespace.RootContext(testCtx), req)
+			if err != nil {
+				t.Fatalf("adding a same-named plugin of a different type should be no problem: %s", err)
+			}
+
+			// Sleep a bit before cleanup is called
+			time.Sleep(1 * time.Second)
+		})
+	}
+}
+
+func TestSystemBackend_Plugin_CatalogRemoved(t *testing.T) {
+	t.Run("secret", func(t *testing.T) {
+		t.Parallel()
+		testPlugin_CatalogRemoved(t, logical.TypeLogical, false, logicalVersionMap)
+	})
+
+	t.Run("auth", func(t *testing.T) {
+		t.Parallel()
+		testPlugin_CatalogRemoved(t, logical.TypeCredential, false, credentialVersionMap)
+	})
+
+	t.Run("secret-mount-existing", func(t *testing.T) {
+		t.Parallel()
+		testPlugin_CatalogRemoved(t, logical.TypeLogical, true, logicalVersionMap)
+	})
+
+	t.Run("auth-mount-existing", func(t *testing.T) {
+		t.Parallel()
+		testPlugin_CatalogRemoved(t, logical.TypeCredential, true, credentialVersionMap)
+	})
+}
+
+func testPlugin_CatalogRemoved(t *testing.T, btype logical.BackendType, testMount bool, versionMap map[string]string) {
+	testCases := []struct {
```

Comment:
I see this is copy/paste, but should we mark this as a helper? same goes for other `testPlugin` functions below

Reply:
I guess since these are running the tests it's not necessary. Sorry for the noise.